File: audio.py
#!/usr/bin/env python

import logging
import math
import numpy as np
import pyaudio
import socket
import struct
import time

# Consts
SAMPLE_RATE = 44100
CHANNELS = 1
FORMAT = pyaudio.paInt16
PCM_QUANT = 32767.5

# Returns nsamples samples of zero
def silence(nsamples):
  samples = [0] * nsamples
  return np.array(samples)

# Returns a sine wave at freq Hz for nsamples samples, varying between +/-1.0
def sinewave(freq, nsamples):
  twoPiFOnRate = 2 * math.pi * freq / float(SAMPLE_RATE)
  samples = []
  for i in xrange(nsamples):
      samples.append(math.sin(i * twoPiFOnRate))
  return np.array(samples)

def sinewaves(freqs, nsamples):
  #begin = time.time()
  # Direct sum of sinewaves
  #waveforms = [ sinewave(f, nsamples) for f in freqs ]
  #waveform = combine(waveforms)
  #print time.time() - begin
  #return waveform

  # Or, inverse FFT, ~10x faster
  # A zero-frequency bucket, then nsamples/2 frequency buckets
  buckets = [0] * (nsamples / 2 + 1)
  for f in freqs:
    bucket = f / float(SAMPLE_RATE) * nsamples
    #assert bucket == int(bucket)
    buckets[int(bucket)] = 1.0
  waveform = np.fft.irfft(buckets, nsamples)
  # Normalize
  waveform = waveform / max(waveform)
  #print time.time() - begin
  return waveform

# Sums and normalizes a collection of waveforms to +/- 1.0
def combine(waveforms):
  agg = sum(waveforms)
  # TODO: implement dynamic range compression instead
  agg = normalise(agg)
  return agg

def normalise(waveform):
  return waveform / max(abs(waveform))

def limit(waveform):
  for i in xrange(len(waveform)):
    waveform[i] = min(waveform[i], 1.0)
    waveform[i] = max(waveform[i], -1.0)

# Compute a signal with energy at each frequency corresponding to a
# non-zero number in chip
def buildWaveform(chip, base, spacing, samples):
  freqs = [ base + spacing * i for i in xrange(len(chip)) if chip[i] > 0 ]
  return sinewaves(freqs, samples)

# Envelopes a waveform by a window function
def window(waveform):
    return np.hanning(len(waveform)) * waveform

# Fades a waveform in-place
def fadein(waveform, samples):
  n = min(len(waveform), samples)
  for i in xrange(samples):
    waveform[i] *= (float(i) / samples)

# Fades a waveform in-place
def fadeout(waveform, samples):
  n = min(len(waveform), samples)
  for i in xrange(len(waveform) - n, len(waveform)):
    waveform[i] *= (len(waveform) - i) / float(n)

# Encodes waveform amplitudes as a little-endian PCM-16 stream
def encodePcm(waveform):
  shorts = [math.floor(PCM_QUANT * f) for f in waveform]
  return struct.pack("<%dh" % len(shorts), *shorts)

# Decodes a stream of little-endian PCM-16 into waveform amplitudes
def decodePcm(block):
  shorts = struct.unpack("<%dh" % (len(block) / 2), block)
  return (np.array(shorts) + 0.5) / PCM_QUANT

# Computes DFT over an array of time-domain samples
# BucketWidth is the 
def fouriate(samples):
  # rfft computes FFT of real-valued inputs, producing n/2+1 amplitude values.
  # buckets[0] is the zero-frequency term (i.e. signal mean/DC offset)
  buckets = np.fft.rfft(np.array(samples))
  bucketWidth = SAMPLE_RATE / len(samples)
  return Spectrum(buckets, bucketWidth)


class Spectrum(object):
  def __init__(self, data, bucketWidth):
    assert bucketWidth > 0
    self.data = data
    self.bucketWidth = bucketWidth

  # The amplitude of the signal around freq
  def amplitude(self, freq):
    return abs(self.data[self.getIndex(freq)])

  # The power of the signal around freq
  def power(self, freq):
    return self.amplitude(freq)**2

  # The complex-valued FFT value around freq
  def getIndex(self, freq):
    assert freq % self.bucketWidth == 0, '%fHz is not a bucket centre' % freq
    index = freq / self.bucketWidth
    return index

def createAudioStream(pa, isInput):
  return pa.open(
      rate = SAMPLE_RATE,
      channels = CHANNELS,
      input = isInput,
      output = (not isInput),
      format = FORMAT)


# Implied:
#
# interface Receiver {
#   double[] receiveBlock(int numSamples)
# }
#
# interface Sender {
#   void sendBlock(double[] waveForm)
# }
#

# Receiver which reads from PyAudio's default input device
class PyAudioReceiver(object):
  def __init__(self):
    self.stream = createAudioStream(pyaudio.PyAudio(), isInput=True)

  def receiveBlock(self, numSamples):
    block = None
    while block is None:
      try:
        #print self.stream.get_read_available(), "frames available"
        block = self.stream.read(numSamples)
        #print "Read", numSamples, "samples"
      except IOError as ex:
        if ex[1] != pyaudio.paInputOverflowed:
          raise
        logging.info("Dropped audio frame attempting to read %d samples! Input overflowed." % numSamples)

    assert len(block) / 2 == numSamples
    return decodePcm(block)

# Receiver which reads from an input stream
class StreamReceiver(object):
  def __init__(self, stream):
    self.stream = stream

  def receiveBlock(self, numSamples):
    nBytes = numSamples * 2
    block = ""
    while len(block) < nBytes:
      block = block + self.stream.read(nBytes - len(block))
      if len(block) == 0:
        assert False
    return decodePcm(block)


class PyAudioSender(object):
  def __init__(self):
    self.stream = createAudioStream(
      pyaudio.PyAudio(), isInput=False)

  def sendBlock(self, waveform):
    self.stream.write(encodePcm(waveform))

# Sender which writes to an output stream
class StreamSender(object):
  def __init__(self, stream):
    self.stream = stream

  def sendBlock(self, waveform):
    self.stream.write(encodePcm(waveform))
    self.stream.flush()

File: native/audio.cpp
#include "audio.h"

#include <cassert>
#include <cmath>
#include <iostream>
#include <vector>

#include "constants.h"
#include "spectrum.h"
#include "util.h"
#include "log.h"
#include "kiss_fftr.h"

#include "stream.h"

using namespace std;

static const float PCM_QUANT = 32767.5;

void silence(int nsamples, vector<float> &target) {
  target.insert(target.end(), nsamples, 0);
}

void sinewave(float freq, int nsamples, vector<float> &target) {
  target.reserve(target.size() + nsamples);
  vector<float>::iterator tit = target.end();
  float twoPiOnRate = 2.0 * M_PI * freq / float(SAMPLE_RATE);
  for (int i = 0; i < nsamples; ++i) {
    target.insert(tit, sin(i * twoPiOnRate));
    ++tit;
  }
}

void sinewaves(vector<float> &frequencies, int nsamples, vector<float> &target) {
  //cerr << "IFFT for " << frequencies << endl;
  kiss_fft_cpx buckets[nsamples / 2 + 1];
  bzero(buckets, sizeof(buckets));
  for (vector<float>::iterator it = frequencies.begin(); it != frequencies.end(); ++it) {
    int bucket = (int)((*it / SAMPLE_RATE) * nsamples);
    buckets[bucket].r = 1.0;
    buckets[bucket].i = 0.0;
   //cerr << bucket << ", ";
  }
  //cerr << endl;

  vector<float> intermediate(nsamples);

  // Allocate FFT memory on stack
  size_t fftMemNeeded = 0;
  kiss_fftr_alloc(nsamples, 1, 0, &fftMemNeeded);
  assert(fftMemNeeded > 0);
  char fftBuffer[fftMemNeeded];

  kiss_fftr_cfg ifft = kiss_fftr_alloc(nsamples, 1, fftBuffer, &fftMemNeeded);
  assert(ifft == (void*)fftBuffer);
  kiss_fftri(ifft, buckets, intermediate.data());

  normalize(intermediate);
  //cerr << intermediate << endl;

  // For testing
  //Spectrum s(intermediate.data(), SAMPLE_RATE, nsamples);
  //for (int i = 300; i < 316; i += 2) {
  //  cerr << s.at(i) << ", abs " << s.amplitude(i) << endl;
  //}
  //cerr << "---" << endl;

  target.insert(target.end(), intermediate.begin(), intermediate.end());
}

void normalize(vector<float> &waveform) {
  normalize(waveform.begin(), waveform.end());
}

void normalize(vector<float>::iterator begin, vector<float>::iterator end) {
  float divisor = 0.0;
  vector<float>::iterator it;
  for (it = begin; it != end; ++it) {
    float d = abs(*it);
    if (d > divisor) { divisor = d; }
  }

  if (divisor > 0.0) {
    for (it = begin; it != end; ++it) {
      *it = *it / divisor;
    }
  }
}

void limit(vector<float> &waveform) {
  vector<float>::iterator it;
  for (it = waveform.begin(); it != waveform.end(); ++it) {
    if (*it > 1.0) { *it = 1.0; }
    if (*it < -1.0) { *it = -1.0; }
  }
}

void buildWaveform(vector<bool> &chip, float base, float channelSpacing, int nsamples, 
    vector<float> &target) {
  vector<float> freqs;
  for (int i = 0; i < chip.size(); ++i) {
    if (chip[i]) {
      freqs.push_back(base + channelSpacing * i);
    }
  }
  sinewaves(freqs, nsamples, target);
  fadein(target.end() - nsamples, nsamples / 10);
  fadeout(target.end() - nsamples/10, nsamples / 10);
}

void window(vector<float> &waveform) {
  float constant = 2.0 * M_PI / (waveform.size() - 1);
  for (int i = 0; i < waveform.size(); ++i) {
    waveform[i] = waveform[i] * 0.5 * (1.0 - cos(i * constant));
  }
}

void fadein(vector<float>::iterator it, int nsamples) {
  for (int i = 0; i < nsamples; ++i) {
    *it *= ((float)i) / nsamples;
    it++;
  }
}

void fadeout(vector<float>::iterator it, int nsamples) {
  for (int i = 0; i < nsamples; ++i) {
    *it *= (nsamples - (float)i) / nsamples;
    it++;
  }
}

void decodePcm16(char *buffer, int buflen, vector<float> &target) {
  target.reserve(target.size() + buflen / 2);
  for (int i = 0; i < buflen; i += 2) {
    short *ps = (short *)(buffer + i);
    target.push_back(((float)*ps + 0.5f) / PCM_QUANT);
  }
}

void encodePcm16(vector<float> &samples, char *buffer) {
  for (int i = 0; i < samples.size(); ++i) {
    short s = floor(PCM_QUANT * samples[i]);
    ((short *)buffer)[i] = s;
  }
}

void PcmDecoder::doPull() {
  assert(source.size() % 2 == 0);

  for (int i = 0; i < source.size(); i += 2) {
    short *ps = (short *)(source.raw() + i);
    push_back(((float)*ps + 0.5f) / PCM_QUANT);
  }

  source.consume(source.size());
}
